Correctly find a dSYM for a file with dots in the filename and using target.exec-search-paths

jasonmolenda opened the pull request:
When lldb is debugging a process on a remote system, it may need to look in the target.exec-search-paths setting to find a binary on the remote host which is not present on the debug host system. If the binary is in a Darwin bundle, like com.apple.sbd.xpc/com.apple.sbd, and lldb is looking for a binary called "com.apple.sbd", it needs to know to (1) look for "com.apple.sbd.xpc" in the target.exec-search-paths, and (2) to look for "com.apple.sbd.xpc.dSYM" next to that bundle to get the debug information correctly.

The old algorithm for testing the parent directories was using the dot's in the filenames to strip off filename components.

rdar://problem/31825940

jasonmolenda committed:
Move the code to find a binary in a bundle in the target.exec-search-paths
from PlatformRemoteDarwinDevice into PlatformDarwin, and have both
PlatformRemoteDarwinDevice and PlatformMacOSX call it there.
<rdar://problem/31825940>

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@304520 91177308-0d34-0410-b5e6-96231b3b80d8
(cherry picked from commit 6b39db95faf7d8458438962758a5f1874b4717d8)
(cherry picked from commit 9e0efee1790a12d170e48e395ae1133a8027f5f0)

jasonmolenda committed:
Change how LocateDSYMInVincinityOfExecutable strips off path
components to not depend on "." characters in the fileanme
(e.g. "Foundation.framework") but instead to just use path
separators.  The names of the files themselves may have dots
in them ("com.apple.sbd") which would break the old scheme.

Also add a test case for this (macosx/find-dsym/bundle-with-dot-in-filename)
as well as a test case for r304520 (macosx/find-dsym/deep-bundle)
which needed a similar setup to test correctly on a single machine.
(both of these are really testing remote debug session situations
where the binary can't be found on the system where lldb is running,
complicating the test case a bit.)

<rdar://problem/31825940>

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@305441 91177308-0d34-0410-b5e6-96231b3b80d8
(cherry picked from commit 518c16c32b266663614c227575537c3e75f88784)
(cherry picked from commit 063b48b4a2e09d5a12e448b9072d08b5edb7e11d)

jasonmolenda commented:
@swift-ci please test

scallanan commented on source/Host/common/Symbols.cpp:
```diff
@@ -104,38 +103,41 @@ static bool LocateDSYMInVincinityOfExecutable(const ModuleSpec &module_spec,
+          // bundle or we've stripped off enough path components that there's no
+          // need to continue.
+
+          for (int i = 0; i < 4; i++) {
```
It looks to me like all these operations are supported by `FileSpec`, e.g. `CopyByRemovingLastPathComponent` for your `rfind`-based approach and `AppendPathComponent` instead of your `+=`.  In fact, it looks like the FindBundleBinaryInExecSearchPaths uses these APIs.  Can we avoid doing this as `std::string` processing?

jasonmolenda replied:
Will look at using a FileSpec instead, thanks.

scallanan commented on source/Plugins/Platform/MacOSX/PlatformDarwin.cpp:
```diff
@@ -1945,3 +1946,79 @@ PlatformDarwin::LaunchProcess(lldb_private::ProcessLaunchInfo &launch_info) {
 }
+
+lldb_private::Error
+PlatformDarwin::FindBundleBinaryInExecSearchPaths (const ModuleSpec &module_spec, Process *process, 
```
This is largely copied from `PlatformRemoteDarwinDevice.cpp`.  It looks like we're just making sure `PlatformMacOSX` has access to it too; is that correct?  If so, I'm not going to review this logic closely.

jasonmolenda replied:
Yeah, exactly.

scallanan reviewed (changes requested):
I have some concerns about the `std::string` implementation style that I mentioned inline below.

I'm also wondering whether this will hold up for sub-frameworks.  A cursory search of the sdk using
```
find . -name "*.framework"
```
shows this pattern commonly used.  4 levels won't be enough there; I suggest repeating until `.framework` no longer appears in your path.

jasonmolenda commented:
re. embedded frameworks, that's not a problem.  If you're debugging /S/L/F/Foundation.framework/Versions/A/Frameworks/OtherStuff.framework/OtherStuff, the dSYM will be a sibling of OtherStuff.framework, you don't need to ascend up to the containing bundle.  Also there's no guarantee what the bundle extension will be -- some kext bundles have embedded ("PlugIns") kexts as well.  I haven't seen any Darwin bundle that required going up more than 4 levels.

jasonmolenda committed:
Change the code I added to LocateDSYMInVincinityOfExecutable to use
the FileSpec methods for adding/removing file path components instead
of using std::strings; feedback from Sean on the change I added in
r305441.
<rdar://problem/31825940>

git-svn-id: https://llvm.org/svn/llvm-project/lldb/trunk@305547 91177308-0d34-0410-b5e6-96231b3b80d8
(cherry picked from commit 4b8d7de76594bab50c81345f8a2e578bb237f266)
(cherry picked from commit b29c3b167972b3f1c0e394239d73c816c2799905)

jasonmolenda commented:
@swift-ci please test

scallanan reviewed (approved):
Thanks, that looks a lot cleaner.  And thanks for the explanation about .dSYM locations for nested frameworks.

clayborg commented:
Was this first done in in SVN top of tree? 

> On Jun 16, 2017, at 9:35 AM, Sean Callanan <notifications@github.com> wrote:
> 
> @scallanan approved this pull request.
> 
> Thanks, that looks a lot cleaner. And thanks for the explanation about .dSYM locations for nested frameworks.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub <https://github.com/apple/swift-lldb/pull/209#pullrequestreview-44607098>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEN5baEzjNq2g53qyCMy4MdcM2gEkjZMks5sEq7AgaJpZM4N6tDw>.
>

jasonmolenda commented:
Yeah this is r304520 + r305441 + r305547.

> On Jun 16, 2017, at 9:49 AM, clayborg <notifications@github.com> wrote:
> 
> Was this first done in in SVN top of tree? 
> 
> > On Jun 16, 2017, at 9:35 AM, Sean Callanan <notifications@github.com> wrote:
> > 
> > @scallanan approved this pull request.
> > 
> > Thanks, that looks a lot cleaner. And thanks for the explanation about .dSYM locations for nested frameworks.
> > 
> > —
> > You are receiving this because you are subscribed to this thread.
> > Reply to this email directly, view it on GitHub <https://github.com/apple/swift-lldb/pull/209#pullrequestreview-44607098>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEN5baEzjNq2g53qyCMy4MdcM2gEkjZMks5sEq7AgaJpZM4N6tDw>.
> > 
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jasonmolenda commented:
@swift-ci please test
